### PR TITLE
experimental llama 2 chat support

### PR DESCRIPTION
**Pull request by jphme**
Heavily based on my Vicuna 1.3 implementation in #265 and leaning on FastChats implementation of Llama2 Chat tokenization.

Still experimental and untested, happy for feedback/improvements (I am not sure whether the eos/bos token handling is 100% correct here...).

### CODE REVIEW

**Comment by winglian** [comment]
thanks! would you mind adding some unit tests? for example for sharegpt/vicuna we use some fixtures https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/tests/test_prompt_tokenizers.py#L39-L61

**Comment by jphme** [comment]
> thanks! would you mind adding some unit tests? for example for sharegpt/vicuna we use some fixtures https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/tests/test_prompt_tokenizers.py#L39-L61

Hey @winglian thank you very much. I wanted to finish yersterday but was too tired in the end - want to get everything in a working state (and this was a lot of debugging, as even the original LLAMA implementation and the HF transformers implementation had different tokenization, see .e.g. https://github.com/huggingface/transformers/pull/24935 .

Attention Masks are still off by one in some cases (working on it now), but eveything else should be consistent with the reference implementation now.

If everything works and I did a successful finetune, I will try to add some fixtures. 
A little problem is that I currently have no GPU locally or on a fixed dev server and thus testing is always difficult outside of finetuning runs on Lambdalabs/runpod (and I am still quite new so don't really have a good setup with sufficient automation).

**Comment by jphme** [comment]
Should be ready to use now, I double checked the tokenization with [Transformers](https://github.com/huggingface/transformers/blob/7381987f906114a3b3c1297f2954df5a63c3eb69/src/transformers/models/llama/tokenization_llama.py#L334) and the original Llama implementation and spend a lot of time to be sure that the masking works correctly.

Will try to add tests/fixtures but help would be appreciated

**Comment by jphme** [comment]
added fixtures now (and even an additional test that compares to transformers tokenization. Unfortunately they don't implement attention masking, otherwise we could've just used their method....).

**Comment by jphme** [comment]
somehow the tokenized file messed up pre-commits again? Should be fixed (and my finetune with this implementation is running now...)

**Review by winglian — APPROVED** [review]

**Comment by minosvasilias** [comment]
Why is this using `human` and `gpt` roles instead of the `user` and `assistant` ones outlined in the [original Llama repo](https://github.com/facebookresearch/llama/blob/main/example_chat_completion.py)?

**Comment by jphme** [comment]
> Why is this using `human` and `gpt` roles instead of the `user` and `assistant` ones outlined in the [original Llama repo](https://github.com/facebookresearch/llama/blob/main/example_chat_completion.py)?

To keep it consistent with the ShareGPT and Vicuna training format and make it more convenient to finetune different models without changing the (conversation) data format every time.
From my own experience, its way more convenient to have some kind of standard data format when finetuning (because you will often try to finetune different models with the same data). The Prompt Strategy should be the layer that customizes that standard conversation format for the model architecture. Imho there are different trade-offs for tuning and inference...

Changing that is also trivial (just adapt the prompter to the different role names) but changing the role names in your dataset is even more trivial.

(But that all my personal opinion, happy to change that if people finde it confusing)

**Comment by minosvasilias** [comment]
Makes sense, thanks for explaining.
If the role names were retained/injected into the prompt, it'd be a different story, but since it's just SYS and INST tokens it makes no practical difference and i agree consistency across models is nice. 👍

**Comment by tmm1** [comment]
This is great, thanks @jphme!
